lib/python/rosie/ws.py

```diff
@@ -189,33 +239,203 @@ def _render(self, all_revs=0, data=None, filters=None, s=None):
+
+    HELLO = "Hello %s\n"
+
+    def get(self):
+        """Say Hello on success."""
+        data = self.HELLO % pwd.getpwuid(os.getuid()).pw_name
+        if self.get_query_argument("format", default=None) == "json":
+            self.write(json.dumps(data))
+        else:
+            self.write(data)
+
+
+class SearchHandler(RosieDiscoService):
+
+    """Serves a search of the database on the page of a given prefix."""
+
+    def get(self, _, all_revs=0):
+        """Search database for rows with data matching the search string."""
+        all_revs = int(all_revs)
+        # None default produces a blank search i.e. returns all suites in DB.
+        s_arg = self.get_query_argument("s", default=None)
+        data = self.dao.search(s_arg, all_revs)
+        if self.get_query_argument("format", default=None) == "json":
+            self.write(json.dumps(data))
+        else:
+            self._render(all_revs, data, s=s_arg)
+
+
+class QueryHandler(RosieDiscoService):
+
+    """Serves a query of the database on the page of a given prefix."""
+
+    def get(self, _, all_revs=0):
+        """Search database for rows with data matching the query string."""
+        all_revs = int(all_revs)
+        q_args = self.get_query_arguments("q")
+        filters = []
+        if not isinstance(q_args, list):
+            q_args = [q_args]
+        filters = [self._query_parse_string(q_str) for q_str in q_args]
+        data = self.dao.query(filters, all_revs)
+        if self.get_query_argument("format", default=None) == "json":
+            self.write(json.dumps(data))
+        else:
+            self._render(all_revs, data, filters=filters)
+
+    @staticmethod
+    def _query_parse_string(q_str):
+        """Split a query filter string into component parts."""
+        conjunction, tail = q_str.split(" ", 1)
+        if conjunction == "or" or conjunction == "and":
+            q_str = tail
+        else:
+            conjunction = "and"
+        filt = [conjunction]
+        if all(s == "(" for s in q_str.split(" ", 1)[0]):
+            start_group, q_str = q_str.split(" ", 1)
+            filt.append(start_group)
+        key, operator, value = q_str.split(" ", 2)
+        filt.extend([key, operator])
+        last_groups = value.rsplit(" ", 1)
+        if (len(last_groups) > 1 and last_groups[1] and
+                all([s == ")" for s in last_groups[1]])):
+            filt.extend(last_groups)
+        else:
+            filt.extend([value])
+        return filt
+
+
+def _log_server_status(application, host, port):
+    """ Log a brief status, including process ID, for an application server."""
+    log_root = os.path.expanduser(LOG_ROOT_TMPL % {
+        "ns": application.NAMESPACE,
+        "util": application.UTIL,
+        "host": host,
+        "port": port})
+    log_status = log_root + ".status"
+    if not os.path.isdir(os.path.dirname(log_root)):
+        os.makedirs(os.path.dirname(log_root))
```

Comment:
Use `os.makedirs(os.path.dirname(log_root), exist_ok=True)` - and remove the `if not ...` test.

Reply:
Great points. Thanks Matt, I'll get them implemented.